### PR TITLE
[runtime env] Put runtime env into runtime context;

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?

Add runtime env into runtime context so that driver/client will be able to retrieve runtime env in an easier way.
## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/protobuf/common.proto [hunk 1]
```diff
@@ -138,7 +138,11 @@ message RayException {
 /// The runtime environment describes all the runtime packages needed to
 /// run some task or actor.
 message RuntimeEnv {
-  repeated string uris = 1;
+  /// The hashed digest of the runtime env.
+  /// It's supposed to be the unique id for a given runtime env
+  string digest_id = 1;
```
Comment (iycheng) [inline src/ray/protobuf/common.proto:143]: I put a specific `digest_id` here in case the id can't be generated by URIs later in the future.

Comment (wuisawesome) [reply src/ray/protobuf/common.proto:143]: Can we specify (in the comment) that this is the md5 of the uploaded uris? (as opposed to the serialized runtime env or something)

File: python/ray/util/client/server/server.py [hunk 2]
```diff
@@ -155,6 +155,7 @@ def ClusterInfo(self, request,
                 ctx.node_id = rtc.node_id.binary()
                 ctx.capture_client_tasks = \
                     rtc.should_capture_child_tasks_in_placement_group
+                ctx.runtime_env = pickle.dumps(rtc.runtime_env)
```
Comment (ericl) [inline python/ray/util/client/server/server.py:158]: Any reason to use pickle instead of JSON?

File: python/ray/util/client/runtime_context.py [hunk 3]
```diff
@@ -41,3 +42,7 @@ def current_node_id(self) -> "NodeID":
     @property
     def should_capture_child_tasks_in_placement_group(self) -> bool:
         return self._fetch_runtime_context().capture_client_tasks
+
+    @property
+    def runtime_env(self) -> Dict:
+        return pickle.loads(self._fetch_runtime_context().runtime_env)
```
Comment (ericl) [inline python/ray/util/client/runtime_context.py:48]: `json.loads`?

File: python/ray/util/client/server/server.py (hunk 2 above)
Comment (iycheng) [reply python/ray/util/client/server/server.py:158]: I'm thinking that right now it's json, but later we probably want to use RuntimeEnvDict.

I'm ok with `json.dump` if you like.

**Comment by wuisawesome** [comment]
Can we just use the `json.loads(worker.core_worker.get_job_config().serialized_runtime_env)` for now? In the common case (small or empty runtime env) there's very little overhead and we don't have to worry about garbage collecting this value from internal kv. 

If this causes a performance regression with big runtime environments, we can rethink it later right?

**Comment by iycheng** [comment]
> Can we just use the `json.loads(worker.core_worker.get_job_config().serialized_runtime_env)` for now? In the common case (small or empty runtime env) there's very little overhead and we don't have to worry about garbage collecting this value from internal kv.
> 
> If this causes a performance regression with big runtime environments, we can rethink it later right?

I'm OK with this. The only concern is that this string will be carried around and it'll also be stored in GCS per-detached actor.

**Comment by wuisawesome** [comment]
But that's already the case and this PR doesn't remove that field, so I feel like we may as well take the simple approach. 

Moving the serialized runtime env to internal kv w/ reference counting seems like a reasonable logical "unit" for a PR in the future.

**Comment by ericl** [comment]
+1 to assuming the string is small and just copying it for each actor

Eric

**Comment by iycheng** [comment]
@wuisawesome it's still different. For detached actors, they'll duplicate the copy. And for every worker, it'll carry the fields around. In kv, we'll only have one copy, but put it in this field, we'll have multiple copy.

But I'm ok with this approach since it's cleaner and later anyway, we'll unify it with RuntimeEnvDict which has already been passed to tasks through cmd.

**Review by wuisawesome — APPROVED** [review]

**Comment by wuisawesome** [comment]
Travis flakey tests look unrelated. Merging.

**Pull request merged**